test/e2e/lib/metrics.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lib
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"knative.dev/pkg/test/monitoring"
+)
+
+// printAllPodMetricsIfTestFailed lists all Pods in the namespace and attempts to print out their
+// metrics.
+//
+// The hope is that this will allow better understanding of where events are lost. E.g. did the
+// source try to send the event to the channel/broker/sink?
+func printAllPodMetricsIfTestFailed(client *Client) {
+	if !client.T.Failed() {
+		// No failure, so no need for logs!
+		return
+	}
+	pods, err := client.Core.Kube.Kube.CoreV1().Pods(client.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		client.T.Logf("Unable to list pods: %v", err)
+		return
+	}
+	for _, pod := range pods.Items {
+		printPodMetrics(client, pod)
+	}
+}
+
+// printPodMetrics attempts to print the metrics from a single Pod to the test logs.
+func printPodMetrics(client *Client, pod corev1.Pod) {
+	podName := types.NamespacedName{
+		Namespace: pod.Namespace,
+		Name:      pod.Name,
+	}
+
+	metricsPort := -1
+	for _, c := range pod.Spec.Containers {
+		for _, p := range c.Ports {
+			if p.Name == "metrics" {
+				metricsPort = int(p.ContainerPort)
+				break
+			}
+		}
+	}
+	if metricsPort < 0 {
+		client.T.Logf("Pod '%v' does not have a metrics port", podName)
+		return
+	}
+
+	root, err := getRootOwnerOfPod(client, pod)
+	if err != nil {
+		client.T.Logf("Unable to get root owner of the Pod '%v': %v", podName, err)
+		root = "root-unknown"
+	}
+
+	podList := &corev1.PodList{
+		Items: []corev1.Pod{
+			pod,
+		},
+	}
+	localPort, err := findAvailablePort()
+	if err != nil {
+		client.T.Logf("Unable to find an avaiable port for Pod '%v': %v", podName, err)
+		return
+	}
+	// There is almost certainly a better way to do this, but for now, just use kubectl to port
```

Comment:
What would be a better way to do this? Can we add an issue for further tracking? I would be interested in learning more :)

Reply:
I don't know either :P 

It just seems like there _should_ be an easier way to do this. If nothing else, a programmatic call to a library that does this, rather than spawning a new process would be preferred. As I was forcing this to fail locally to make sure things worked, I accidentally orphaned a few of these port forward processes. If it was an in-process library call that wouldn't be possible. It also would allow me to remove the polling loop waiting for the port forwarding to actually start working. And I would know if something went wrong (as opposed to today where I never read the output of the executed kubectl call).

Here is a blog post that has a promising title (haven't actually read it yet): https://gianarb.it/blog/programmatically-kube-port-forward-in-go

Filed https://github.com/google/knative-gcp/issues/1083. This should also be done in knative/pkg, should I make an issue there too?